admin-dev/themes/new-theme/js/components/form/form-object-mapper.js

```diff
@@ -0,0 +1,282 @@
+ */
+
+import _ from 'lodash';
+
+const {$} = window;
+
+/**
+ * This is able to watch an HTML form and parse it as a Javascript object based on a configurable
+ * mapping. Each field from the model is mapped to a form input, or several, each input is watched
+ * to keep the model consistent.
+ *
+ * When a model field has several inputs associated it keeps the in sync when either of them is updated.
+ */
+export default class FormObjectMapper {
+  /**
+   * @param $form {jQuery}
+   * @param modelMapping {Object}
+   * @param eventEmitter {EventEmitter}
+   * @param config {Object}
+   */
+  constructor($form, modelMapping, eventEmitter, config) {
+    this.$form = $form;
+    this.fullModelMapping = modelMapping;
+    this.eventEmitter = eventEmitter;
+
+    const inputConfig = config || {};
+
+    // This event is registered so when it is triggered it forces the form mapping and object update,
+    // it can be useful when some new inputs have been added in the DOM (or removed) so that the model
+    // acknowledges the update
+    this.updateModelEventName = inputConfig.updateModel || 'updateModel';
+
+    // This event is emitted each time the object is updated (from both input change and external event)
+    this.modelUpdatedEventName = inputConfig.modelUpdated || 'modelUpdated';
+    // This event is emitted each time an object field is updated (from both input change and external event)
+    this.modelFieldUpdatedEventName = inputConfig.modelFieldUpdated || 'modelFieldUpdated';
+
+    this.modelMapping = {};
+    this.formMapping = {};
+    this.object = {};
+
+    this.initFormMapping();
+    this.watchUpdates();
+    this.updateFullObject();
+  }
+
+  /**
+   * Returns the object mapped to the form (current live state)
+   *
+   * @returns {*|{}}
+   */
+  getObject() {
+    return this.object;
+  }
+
+  /**
+   * Get a field from the object based on the model key
+   *
+   * @param modelKey (string)
+   *
+   * @returns {*|{}}
+   */
+  get(modelKey) {
+    const modelKeys = modelKey.split('.');
+
+    return $.serializeJSON.deepGet(this.object, modelKeys);
+  }
+
+  /**
+   * Returns the input associated to a model field (in case the field is mapped to
+   * several inputs the default one, first configured, is always used).
+   *
+   * @param modelKey {string}
+   * @returns {undefined|jQuery}
+   */
+  getInput(modelKey) {
+    if (!Object.prototype.hasOwnProperty.call(this.modelMapping, modelKey)) {
+      return undefined;
+    }
+
+    return $(`[name="${this.modelMapping[modelKey]}"]`);
+  }
+
+  /**
+   * Set a value to a field of the object based on the model key, the object itself is updated
+   * of course but the mapped inputs are also synced (all of them if multiple). Events are also
+   * triggered to indicate the object has been updated (the general and the individual field ones).
+   *
+   * @param modelKey {string}
+   * @param value {*|{}}
+   */
+  set(modelKey, value) {
```

Comment:
Same here, I don't know at all what I need to set when using this method, do I need to set an input key, or a data ? Obviously you can guess it while checking the doc or even the signature..

When using PHP, how do you name setters and getters ?